File: scripts/algorithms/Binary Tree Zigzag Level Order Traversal/Binary Tree Zigzag Level Order Traversal.js
var zigzagLevelOrder = function(root) {
    let result = []
    if(root == null) return result
    
    let queue = []
    let leftToRight = true
    
    queue.push(root)
    
    while(queue.length != 0) {
        let qSize = queue.length
        let ans = []
        
        // Processing level
        for(let i=0;i<qSize;i++) {
            let front = queue.shift()
            
            let index = leftToRight ? i : qSize-i-1
            ans[index] = front.val
            
            if(front.left) queue.push(front.left)
            if(front.right) queue.push(front.right)
        }
        // changing order of operation
        leftToRight = !leftToRight
        
        result.push(ans)
    }
    
    return result
};

File: scripts/algorithms/Binary Tree Zigzag Level Order Traversal/Binary Tree Zigzag Level Order Traversal.py
class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root: return []
        
        ans = []
        
        node = root 
        
        q = collections.deque([node])
        
        order = -1 
        
        while q:
            order = -order
            level = []
            for _ in range(len(q)):
                node = q.popleft()
                
                level.append(node.val)
                
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
                    
            ans.append(level[::order])
                
        return ans
